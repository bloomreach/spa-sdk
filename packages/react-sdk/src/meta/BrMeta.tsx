/*
 * Copyright 2019 Hippo B.V. (http://www.onehippo.com)
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import React from 'react';
import ReactDOM from 'react-dom';
import { MetaComment, Meta, META_POSITION_BEGIN, META_POSITION_END, isMetaComment } from '@bloomreach/spa-sdk';

interface BrMetaProps {
  meta: Meta[];
}

export class BrMeta extends React.Component<BrMetaProps> {
  private comments: Comment[] = [];
  private head?: Element | Text;
  private tailRef = React.createRef<HTMLElement>();

  componentDidMount() {
    this.head = ReactDOM.findDOMNode(this)!;
    this.renderMeta();
  }

  componentDidUpdate() {
    this.head = ReactDOM.findDOMNode(this)!;
    this.removeMeta();
    this.renderMeta();
  }

  componentWillUnmount() {
    this.removeMeta();
  }

  private removeMeta() {
    this.comments.splice(0).forEach(comment => comment.remove());
  }

  private renderMeta() {
    this.props.meta.filter(meta => isMetaComment(meta) && meta.getPosition() === META_POSITION_BEGIN)
      .forEach(this.renderMetaComment.bind(this));
    this.props.meta.filter(meta => isMetaComment(meta) && meta.getPosition() === META_POSITION_END)
      .reverse()
      .forEach(this.renderMetaComment.bind(this));
  }

  private renderMetaComment(meta: MetaComment) {
    const comment = this.head!.ownerDocument!.createComment(meta.getData());
    this.comments.push(comment);

    if (meta.getPosition() === META_POSITION_BEGIN) {
      return void this.head!.parentNode!.insertBefore(comment, this.head!);
    }

    if (!this.tailRef.current!.nextSibling) {
      return void this.tailRef.current!.parentNode!.appendChild(comment);
    }

    this.tailRef.current!.parentNode!.insertBefore(comment, this.tailRef.current!.nextSibling);
  }

  render() {
    return (
      <>
        {this.props.children}
        {this.props.meta.length ? <span style={{ display: 'none' }} ref={this.tailRef} /> : null}
      </>
    );
  }
}
