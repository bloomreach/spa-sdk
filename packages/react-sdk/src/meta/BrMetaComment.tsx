/*
 * Copyright 2019 Hippo B.V. (http://www.onehippo.com)
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import React from 'react';
import { MetaComment } from '@bloomreach/spa-sdk';

interface BrMetaCommentProps {
  meta: MetaComment;
}

export class BrMetaComment extends React.Component<BrMetaCommentProps> {
  private placeholder = React.createRef<HTMLElement>();
  private comment?: Comment;

  componentDidMount() {
    const placeholder = this.placeholder.current!;
    this.comment = placeholder.ownerDocument!.createComment(this.props.meta.getData());
    placeholder.parentNode!.replaceChild(this.comment, placeholder);
  }

  shouldComponentUpdate(nextProps: BrMetaCommentProps) {
    return nextProps.meta.getData() !== this.props.meta.getData();
  }

  componentDidUpdate() {
    this.comment!.replaceData(0, this.comment!.data.length, this.props.meta.getData());
  }

  componentWillUnmount() {
    this.comment!.parentNode!.replaceChild(this.placeholder.current!, this.comment!);
    delete this.comment;
  }

  render() {
    return <span style={{ display: 'none' }} ref={this.placeholder} />;
  }
}
