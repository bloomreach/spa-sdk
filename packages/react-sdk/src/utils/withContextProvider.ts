// eslint-disable-next-line import/no-extraneous-dependencies
import PropTypes from 'prop-types';
import React, { ReactElement } from 'react';

export const withContextProvider = (context?: Record<string, any>, children?: ReactElement): React.ReactElement => {
  const childContextTypes = (): Record<string, string> | undefined => context
    && Object.keys(context).reduce((obj: Record<string, any>, key: string) => {
      obj[key] = PropTypes.any;
      return obj;
    }, {});

  class ContextProvider extends React.Component {
    getChildContext(): Record<string, string> | undefined {
      return context;
    }

    render(): React.ReactNode {
      return children;
    }
  }

  (ContextProvider as any).displayName = 'ContextProvider';
  (ContextProvider as any).childContextTypes = childContextTypes();

  return React.createElement(ContextProvider, null);
};
