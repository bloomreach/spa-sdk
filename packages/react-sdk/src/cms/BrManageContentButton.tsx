/*
 * Copyright 2019-2025 Bloomreach
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *   https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import React, { useContext } from 'react';
import { ManageContentButton, TYPE_MANAGE_CONTENT_BUTTON } from '@bloomreach/spa-sdk';
import { BrMeta } from '../meta';
import { BrPageContext } from '../page/BrPageContext';

/**
 * The button component that opens for editing a content.
 */
export function BrManageContentButton(props: ManageContentButton): React.ReactElement | null {
  const page = useContext(BrPageContext);

  return page?.isPreview() ? <BrMeta meta={page.getButton(TYPE_MANAGE_CONTENT_BUTTON, props)} /> : null;
}
