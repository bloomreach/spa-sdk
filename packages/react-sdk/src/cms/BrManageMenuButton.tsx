/*
 * Copyright 2019 Hippo B.V. (http://www.onehippo.com)
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import React from 'react';
import { Menu } from '@bloomreach/spa-sdk';
import { BrPageContext } from '../page';
import { BrMetaWrapper } from '../meta';

interface BrManageMenuButtonProps {
  menu: Menu;
}

export class BrManageMenuButton extends React.Component<BrManageMenuButtonProps, {}> {
  static contextType = BrPageContext;
  context: React.ContextType<typeof BrPageContext>;

  constructor(props: BrManageMenuButtonProps) {
    super(props);
  }

  render() {
    if (!this.context!.isPreview()) {
      return null;
    }

    if (!this.props.menu._meta) {
      return null;
    }

    const meta = this.context!.getMeta(this.props.menu._meta);
    return <BrMetaWrapper meta={meta} />;
  }
}
