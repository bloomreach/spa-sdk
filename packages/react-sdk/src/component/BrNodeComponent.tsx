/*
 * Copyright 2019-2025 Bloomreach
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *   https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import { Component } from '@bloomreach/spa-sdk';
import React, { useContext } from 'react';
import { BrMeta } from '../meta';
import { BrMappingContext } from './BrMappingContext';
import { BrProps } from './BrProps';

export function BrNodeComponent<T extends Component>(props: React.PropsWithChildren<BrProps<T>>): React.ReactElement {
  const context = useContext(BrMappingContext);
  const { component, children } = props;

  const mapping = component && (context[component.getName()] as React.ComponentType<BrProps>);
  const meta = component?.getMeta();

  const content = mapping ? React.createElement(mapping, props) : children;

  return React.createElement(BrMeta, { meta }, content);
}
