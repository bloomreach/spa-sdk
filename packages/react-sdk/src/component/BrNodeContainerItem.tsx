/*
 * Copyright 2019-2025 Bloomreach
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *   https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import React, { useCallback, useContext, useEffect, useState } from 'react';
import { ContainerItem, TYPE_CONTAINER_ITEM_UNDEFINED } from '@bloomreach/spa-sdk';
import { BrContainerItemUndefined } from '../cms';
import { BrProps } from './BrProps';
import { BrMappingContext } from './BrMappingContext';
import { BrMeta } from '../meta';

export function BrNodeContainerItem(
  props: React.PropsWithChildren<BrProps<ContainerItem>>,
): React.ReactElement {
  const context = useContext(BrMappingContext);
  const { component, page, children } = props;
  const [forcedRerenders, setForcedRerenders] = useState(0);

  const onUpdate = useCallback((): void => {
    // Trigger a rerender of this component due to a non-react 3rd party event
    setForcedRerenders((c) => c + 1);
  }, []);

  useEffect(() => {
    component?.on('update', onUpdate);

    return () => {
      component?.off('update', onUpdate);
    };
  }, [component, onUpdate]);

  useEffect(() => {
    // Skip the mount event page sync as BrPage just called it
    if (forcedRerenders > 0) {
      page?.sync();
    }
  }, [page, forcedRerenders]);

  const getMapping = (): React.ComponentType<BrProps> => {
    const type = component?.getType();

    if (type && type in context) {
      return context[type] as React.ComponentType<BrProps>;
    }

    return (
      (context[
        TYPE_CONTAINER_ITEM_UNDEFINED as any
      ] as React.ComponentType<BrProps>) ?? BrContainerItemUndefined
    );
  };

  const mapping = getMapping();
  const meta = component?.getMeta();
  const content = mapping ? React.createElement(mapping, props) : children;

  return React.createElement(BrMeta, { meta }, content);
}
