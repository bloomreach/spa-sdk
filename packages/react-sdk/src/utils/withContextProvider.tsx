/*
 * Copyright 2019-2025 Bloomreach
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *   https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import React, { ReactElement } from 'react';
import { BrPageContext } from '../page';
import { BrComponentContext, BrMappingContext } from '../component';

export const withPageContextProvider = (context?: any, children?: ReactElement): React.ReactElement => {
  if (!context) {
    return children as React.ReactElement;
  }

  return <BrPageContext.Provider value={context}>{children}</BrPageContext.Provider>;
};

export const withMappingContextProvider = (context?: any, children?: ReactElement): React.ReactElement => {
  if (!context) {
    return children as React.ReactElement;
  }

  return (
    <BrMappingContext.Provider value={context}>
      {children}
    </BrMappingContext.Provider>
  );
};

export const withComponentContextProvider = (context?: any, children?: ReactElement): React.ReactElement => {
  if (!context) {
    return children as React.ReactElement;
  }

  return <BrComponentContext.Provider value={context}>{children}</BrComponentContext.Provider>;
};
