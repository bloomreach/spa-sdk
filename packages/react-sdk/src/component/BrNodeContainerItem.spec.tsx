/*
 * Copyright 2019-2022 Bloomreach
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import React from 'react';
import { ContainerItem, MetaCollection, Page, TYPE_CONTAINER_ITEM_UNDEFINED } from '@bloomreach/spa-sdk';
import { act, render } from '@testing-library/react';
import { BrContainerItemUndefined } from '../cms';
import { BrNodeComponent } from './BrNodeComponent';
import { BrNodeContainerItem } from './BrNodeContainerItem';
import { BrMeta } from '../meta';
import { withContextProvider } from '../utils/withContextProvider';

jest.mock('@bloomreach/spa-sdk', () => () => ({ [TYPE_CONTAINER_ITEM_UNDEFINED]: 'StringValue' }));

describe('BrNodeContainerItem', () => {
  const props = {
    component: {
      getType: jest.fn(),
      on: jest.fn(),
      off: jest.fn(),
      getMeta: jest.fn(),
    } as unknown as jest.Mocked<ContainerItem>,
    page: { sync: jest.fn() } as unknown as jest.Mocked<Page>,
  };

  const emptyMeta = {} as MetaCollection;

  beforeEach(() => {
    jest.clearAllMocks();

    props.component.getMeta.mockReturnValue(emptyMeta);
  });

  describe('componentDidMount', () => {
    it('should subscribe for update event on mount', () => {
      render(<BrNodeContainerItem {...props} />);

      expect(props.component.on).toBeCalledWith('update', expect.any(Function));
    });
  });

  describe('componentDidUpdate', () => {
    it('should resubscribe on component update', () => {
      const element = render(<BrNodeContainerItem {...props} />);

      jest.clearAllMocks();
      render(<BrNodeContainerItem component={{ ...props.component }} />, { container: element.container });

      expect(props.component.off).toBeCalledWith('update', expect.any(Function));
      expect(props.component.on).toBeCalledWith('update', expect.any(Function));
    });
  });

  describe('componentWillUnmount', () => {
    it('should unsubscribe from update event on unmount', () => {
      const element = render(<BrNodeContainerItem {...props} />);
      element.unmount();

      expect(props.component.off).toBeCalledWith('update', props.component.on.mock.calls[0][1]);
    });
  });

  describe('getMapping', () => {
    beforeEach(() => {
      (BrNodeContainerItem as any).contextTypes = { [TYPE_CONTAINER_ITEM_UNDEFINED]: () => null };
      delete (BrNodeComponent as Partial<typeof BrNodeComponent>).contextType;
    });

    it('should use container item type for mapping', () => {
      props.component.getType.mockReturnValueOnce('test');
      (BrNodeContainerItem as any).contextTypes = { test: () => null };
      render(
        withContextProvider(
          {
            test: () => <div>Test</div>,
          },
          <BrNodeContainerItem {...props} />,
        ),
      );

      expect(props.component.getType).toBeCalled();
    });

    it('should render undefined container item', () => {
      const element = render(
        <BrNodeContainerItem {...props}>
          <a />
        </BrNodeContainerItem>,
      );

      const nodeMeta = render(
        <BrMeta meta={emptyMeta}>
          <BrContainerItemUndefined {...props}>
            <a />
          </BrContainerItemUndefined>
        </BrMeta>,
      );

      expect(element.container.isEqualNode(nodeMeta.container)).toBe(true);
      expect(element.asFragment()).toMatchSnapshot();
    });

    it('should override undefined container item', () => {
      props.component.getType.mockReturnValueOnce('test');
      const element = render(
        withContextProvider(
          {
            [TYPE_CONTAINER_ITEM_UNDEFINED]: ({ children }: React.PropsWithChildren<typeof props>) => (
              <div>{children}</div>
            ),
          },
          <BrNodeContainerItem {...props}>
            <a />
          </BrNodeContainerItem>,
        ),
      );

      expect(element.asFragment()).toMatchSnapshot();
    });
  });

  describe('onUpdate', () => {
    it('should trigger sync on update event', async () => {
      render(<BrNodeContainerItem {...props} />);

      await act(() => props.component.on.mock.calls[0][1]({}));

      expect(props.page.sync).toBeCalled();
    });
  });
});
