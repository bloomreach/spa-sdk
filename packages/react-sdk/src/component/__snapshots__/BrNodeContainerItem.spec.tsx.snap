// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BrNodeContainerItem getMapping should override undefined container item 1`] = `
<DocumentFragment>
  <div>
    <a />
  </div>
</DocumentFragment>
`;

exports[`BrNodeContainerItem getMapping should render undefined container item 1`] = `
<DocumentFragment>
  <div>
    Component "undefined" is not defined.
  </div>
</DocumentFragment>
`;
